e2e/testcases/reconciler_finalizer_test.go

```diff
@@ -0,0 +1,702 @@
+	deployment1 := rootRepo.Get(deployment1Path)
+	deployment1.SetName(deployment1Name)
+	deployment1.SetNamespace(repoSyncNN.Namespace)
+	rootRepo.Add(deployment1Path, deployment1)
+	rootRepo.CommitAndPush("Adding deployment helloworld-1 to RootSync")
+	nt.WaitForRepoSyncs()
+	nomostest.WaitForCurrentStatus(nt, kinds.Deployment(), deployment1.GetName(), deployment1.GetNamespace())
+
+	// Add deployment-helloworld-2 to RepoSync
+	deployment2Path := nomostest.StructuredNSPath(repoSyncNN.Namespace, "deployment-helloworld-2")
+	nsRepo.Copy("../testdata/deployment-helloworld.yaml", deployment2Path)
+	deployment2 := nsRepo.Get(deployment2Path)
+	deployment2.SetName(deployment2Name)
+	deployment2.SetNamespace(repoSyncNN.Namespace)
+	nsRepo.Add(deployment2Path, deployment2)
+	nsRepo.CommitAndPush("Adding deployment helloworld-2 to RepoSync")
+	nt.WaitForRepoSyncs()
+	nomostest.WaitForCurrentStatus(nt, kinds.Deployment(), deployment2.GetName(), deployment2.GetNamespace())
+
+	nt.T.Log("Enabling RootSync deletion propagation")
+	rootSync := fake.RootSyncObjectV1Beta1(configsync.RootSyncName)
+	err := nt.Get(rootSync.Name, rootSync.Namespace, rootSync)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	core.SetAnnotation(rootSync,
+		metadata.DeletionPropagationPolicyAnnotationKey,
+		string(metadata.DeletionPropagationPolicyForeground))
+	err = nt.Update(rootSync)
+	if err != nil {
+		nt.T.Fatal(err)
+	}
+	nt.T.Cleanup(func() {
+		nt.T.Log("Resetting RootSync deletion propagation")
+		rootSync := fake.RootSyncObjectV1Beta1(configsync.RootSyncName)
+		err := nt.Get(rootSync.Name, rootSync.Namespace, rootSync)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// ValidateMultiRepoDeployments called by WaitForRepoSyncs handles RootSync re-creation
+				return
+			}
+			nt.T.Fatal(err)
+		}
+		core.RemoveAnnotations(rootSync,
+			metadata.DeletionPropagationPolicyAnnotationKey)
+		err = nt.Update(rootSync)
+		if err != nil {
+			nt.T.Fatal(err)
+		}
+	})
+	nomostest.WaitForCurrentStatus(nt, kinds.RootSyncV1Beta1(), rootSync.GetName(), rootSync.GetNamespace())
+	nomostest.WaitForFinalizer(nt, kinds.RootSyncV1Beta1(), rootSync.GetName(), rootSync.GetNamespace(), metadata.ReconcilerFinalizer)
+
+	nt.T.Log("Disabling RepoSync deletion propagation")
+	repoSync := rootRepo.Get(repoSyncPath)
+	annotations := repoSync.GetAnnotations()
+	if _, found := annotations[metadata.DeletionPropagationPolicyAnnotationKey]; found {
+		core.RemoveAnnotations(repoSync, metadata.DeletionPropagationPolicyAnnotationKey)
+		rootRepo.Add(repoSyncPath, repoSync)
+		rootRepo.CommitAndPush("Disabling RepoSync deletion propagation")
+		nt.WaitForRepoSyncs()
+	}
+	nomostest.WaitForCurrentStatus(nt, kinds.RepoSyncV1Beta1(), repoSync.GetName(), repoSync.GetNamespace())
+	nomostest.WaitForMissingFinalizer(nt, kinds.RepoSyncV1Beta1(), repoSync.GetName(), repoSync.GetNamespace(), metadata.ReconcilerFinalizer)
+
+	// Abandon the test namespace, otherwise it will block the finalizer
```

Comment:
Do we need to add a test for the block behavior and validate the RootSync finalizer error?

Reply:
Now that I've added the conditions I could wait for those to show the error and then resolve it by committing a new change.

But now that I think about it, the Finalizer is shutting down the Parser and Remediator. So new commits will be ignored after finalization starts. That's probably a problem... I'm gonna have to think of a way to stop the finalizer on error and restart the other controllers to allow the user to fix the problem. Tho I'm not sure if that needs to be done for this MVP or not, since we only want to use it for test cleanup initially.

Reply:
Correction: 
I think it's ok to shut down the Parser and Remediator. We don't need to re-start them. There should never be a situation where you were able to apply objects but not able to delete them (in reverse dependency order). I think...

By removing the early return in the reconciler-manager reconcile loop when there's a deletionTimestamp, I've made it possible for the user to fix any permission of configuraiton problems with the RSync or user-managed Secrets. This should allow the user to unblock the finalizer, because the finalizer will error or timeout and retry. And even if the reconciler-manager modifies the reconciler Deployement, causing the reconciler Pod to be replaced, the new reconciler will detect the deletionTimestamp and attempt to destroy any managed objects immediately after starting up. 

If there's any edge cases I haven't forseen, we should be able to fix them and instruct the user to delete the reconciler Pod, if it gets wedged. So there's still a workaround if I missed something.